private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,634 @@
+	return completionsToCompletionList(options), nil
+}
+
+func (s *server) Definition(ctx context.Context, params *protocol.DefinitionParams) ([]protocol.Location /* Definition | DefinitionLink[] | null */, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	entry, ok := s.fileCache[params.TextDocument.URI.Filename()]
+	if !ok {
+		return nil, fmt.Errorf("unknown file: %s", params.TextDocument.URI)
+	}
+	pos := entry.getSourcePos(params.Position)
+
+	var result []protocol.Location
+	if importEntry := entry.findImportEntry(pos); importEntry != nil {
+		if importEntry.docURI == "" {
+			return nil, nil
+		}
+		result = []protocol.Location{{URI: importEntry.docURI}}
+	} else {
+		result = s.findReferencedDefLoc(ctx, entry, pos)
+	}
+
+	// Make sure all the files exist.
+	for _, loc := range result {
+		if strings.HasPrefix(loc.URI.Filename(), s.container.CacheDirPath()) {
+			// This is a temporary file, make sure it exists.
+			localPath := loc.URI.Filename()
+			if _, err := os.Stat(localPath); err != nil {
+				tmpEntry, ok := s.fileCache[loc.URI.Filename()]
+				if !ok {
+					return nil, fmt.Errorf("unknown file: %s", loc.URI)
+				}
+
+				// Create the file.
+				if err := os.MkdirAll(path.Dir(localPath), 0755); err != nil {
+					return nil, err
+				}
+				tmpFile, err := os.OpenFile(localPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0444)
+				if err != nil {
+					return nil, err
+				}
+				defer tmpFile.Close()
+				if _, err := tmpFile.WriteString(tmpEntry.document.Text); err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func (s *server) Hover(ctx context.Context, params *protocol.HoverParams) (*protocol.Hover, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	entry, ok := s.fileCache[params.TextDocument.URI.Filename()]
+	if !ok {
+		return nil, fmt.Errorf("unknown file: %s", params.TextDocument.URI)
+	}
+	pos := entry.getSourcePos(params.Position)
+
+	symbols := s.findReferencedSymbols(ctx, entry, pos)
+	if len(symbols) == 0 {
+		return nil, nil
+	}
+	symbol := symbols[0]
+	refEntry, ok := s.fileCache[symbol.file.Filename()]
+	if !ok {
+		return nil, fmt.Errorf("unknown file: %s", symbol.file)
+	}
+
+	codeData, err := refEntry.genNodeSignature(symbol.node)
+	if err != nil {
+		return nil, err
+	}
+
+	return &protocol.Hover{
+		Contents: protocol.MarkupContent{
+			Kind:  protocol.Markdown,
+			Value: fmt.Sprintf("### %s\n```proto\n%s\n```", strings.Join(symbol.name(), "."), codeData),
+		},
+	}, nil
+}
+
+func (s *server) resolveImport(ctx context.Context, workspace bufmodule.ModuleSet, bucket bufmodule.ModuleReadBucket, path string) (*fileEntry, error) {
+	file, err := bucket.GetFile(ctx, path)
+	if err != nil {
+		workspace = s.wellKnownTypesModuleSet
+		bucket = s.wellKnownTypesBucket
+		file, err = bucket.GetFile(ctx, path)
+		if err != nil {
+			s.logger.Warn("could not resolve import", zap.String("path", path))
```

Comment:
Why aren't we returning this error? We generally never want to using logging as a means of error propagation unless absolutely necessary.

Reply:
It is going to be necessary to swallow some errors at some point because the LSP should provide some functionality even when the source code is partially broken, since the LSP will necessarily see source code as it is being typed. In this case, it doesn't actually matter since we bail out earlier in processText anyway, so there's really no point in swallowing this error, so now it does not do that.